api/src/main/java/org/openmrs/module/queue/api/impl/QueueEntryServiceImpl.java

```diff
@@ -100,6 +101,20 @@ public QueueEntry saveQueueEntry(QueueEntry queueEntry) {
 		return dao.createOrUpdate(queueEntry);
 	}
 	
+	/**
+	 * @see QueueEntryService#transitionQueueEntry(QueueEntryTransition)
+	 */
+	@Override
+	public QueueEntry transitionQueueEntry(QueueEntryTransition queueEntryTransition) {
+		// End the initial queue entry
+		QueueEntry queueEntryToStop = queueEntryTransition.getQueueEntryToTransition();
+		queueEntryToStop.setEndedAt(queueEntryTransition.getTransitionDate());
+		saveQueueEntry(queueEntryToStop);
```

Comment:
Unfortunately, the limitation this will bypass the SaveHandlers and any other AOP. The reason is that AOP is implemented (in Spring) using an AOP proxy, but since this results in a direct method call, the proxy can't intercept the call (see [the Spring docs](https://docs.spring.io/spring-framework/docs/5.3.32/reference/html/core.html#aop-understanding-aop-proxies)).

Instead of the call to `Context.getService()` it might be slightly more testable to use Spring's `AopContext`:

```java
((QueueEntryService) AopContext.currentProxy()).saveQueueEntry(queueEntryToStop);
```

This has the advantage that, in test code, you can just write:

```java
AopContext.setCurrentProxy(queueEntryService);
```

Reply:
Ugh, yeah, this is a pain... :)

Reply:
OK.  I'll just make my unit test into a context sensitive one, though that's a cool trick @ibacher .  I had written it the way I was supposed to originally, but then changed when my unit tests failed.  Just hoping maybe this had been magically fixed I guess.

Reply:
Actually I found a different approach that seemed to work out easier.  See last commit